Expand manual token secret creation docs (#33675)
kubernetes/website: content/en/docs/concepts/configuration/secret.md

a service account token Secret; for example, if you need a token that never expires.
However, using the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
subresource to obtain a token to access the API is recommended instead.

> The `kubectl create token` command can be used to obtain a token from the `TokenRequest` API.

Reviewer: Got a link to how to use that command? Or to reference docs?

Author: `kubectl create token --help` has examples

Generated docs are updated in https://github.com/kubernetes/website/pull/33676#issuecomment-1126859819

Reviewer: I mean for the doc. A link to instructions on how to use the command would be useful.

Author: updated with link to the command docs

Reviewer: Thanks!

a service account token Secret; for example, if you need a token that never expires.
However, using the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
subresource to obtain a token to access the API is recommended instead.

> The [`kubectl create token`](/docs/reference/generated/kubectl/kubectl-commands#-em-token-em-)
> command can be used to obtain a token from the `TokenRequest` API.

Reviewer 2: ```suggestion
You can use the [`kubectl create token`](/docs/reference/generated/kubectl/kubectl-commands#-em-token-em-)
command to obtain a token from the `TokenRequest` API.
```

### Service account token Secrets

A `kubernetes.io/service-account-token` type of Secret is used to store a
token credential that identifies a
{{< glossary_tooltip text="service account" term_id="service-account" >}}.

> Since 1.22, this type of Secret is no longer used to mount credentials into Pods,

Reviewer: So in 1.22 & 1.23 it made a secret with token but didn’t use it for the pods?

Author: Yes, to accommodate 2 version node skew, because pods scheduled on 1.21 nodes could still have required secret-based tokens if a cluster admin had disabled bound token projected volumes.

 1.22 is when the secrets injected to pods unconditionally used projected tokens.

and obtaining tokens via the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
API is recommended instead of using service account token Secret objects.
Tokens obtained from the `TokenRequest` API are more secure than ones stored in Secret objects,
because they have a bounded lifetime and are not readable by other API clients.
`kubectl create token` can be used to obtain a token from the `TokenRequest` API.

If you are unable to use the `TokenRequest` API to obtain a token,
and the security exposure of persisting a non-expiring token credential
in a readable API object is acceptable for your use case, 
you can create a service account token Secret object.

When using this Secret type, you need to ensure that the
`kubernetes.io/service-account.name` annotation is set to an existing
service account name. If you are creating both the ServiceAccount and

> the Secret objects, that means you should create the ServiceAccount first.

Reviewer: Does it error if you don’t? Will it work eventually if they apply in the other order?

Author: If you create the secret first it's a race with the service account creation whether the token controller will observe it, check that there's no corresponding service account, and delete it as an orphaned secret.

Reviewer: This is gonna be problematic when generating them with GitOps. 

I don’t think kubectl even applies orders in a specific order when you apply a directory. 

It’s a good thing I’ve put dependency orchestration for kpt & Config Sync, but I hate to see it required for vanilla resources. It’s terrible UX for declarative assets. 

Is there any way we can add another annotation to make the user intent more clear to avoid the race?

Author: > I don’t think kubectl even applies objects in a specific order when you apply a directory.

order is definitely deterministic, and is covered by unit tests to avoid breaking that order: https://github.com/kubernetes/kubernetes/pull/103407

> Is there any way we can add another annotation to make the user intent more clear to avoid the race?

I think requiring the service account to exist is achievable and reasonable. This is similar to setting ownerReferences, where the owning object also has to exist first.

Reviewer: > order is definitely deterministic

By file name, yes, but not resource type or anything more fancy. That's why we added apply ordering by resource type (as well as implicit dependency detection for CRD, Namespaces, and explicit dependencies by annotation) in cli-utils.

> I think requiring the service account to exist is achievable and reasonable. This is similar to setting ownerReferences, where the owning object also has to exist first.

Yes. There are several unfortunate examples of precedence, and all of them have to be accounted for when trying to programmatically handle waiting for reconciliation and apply ordering.

We have a number of exceptions hard coded in the kstatus library: 
https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/kstatus/status/core.go#L22

And there's apply ordering hard coded as well (tho we're planning to replace this with more implicit dependency detection):
https://github.com/kubernetes-sigs/cli-utils/blob/master/pkg/ordering/sort.go#L74

It looks like we haven't been handling ServiceAccount or Secret with any exceptions yet. So they'll need to be added. I'm not sure how to detect if a ServiceAccount is reconciled tho. We weren't previously waiting for the Secret or token to be generated, but that probably means it was considered reconciled immediately, since there's no status.

I'm beginning to think all resources should have status and conditions.

Author: Straight creation in order is sufficient, the token controller will do a live lookup to verify the service account is absent before cleaning up the token, so you don't have to wait after creating the service account to proceed to the secret.

Reviewer: Noted. Adding it to the resource apply order should be sufficient then until we move to async/parallel apply. Then we'll have to add it as an implicit dependency.

### Service account token Secrets

A `kubernetes.io/service-account-token` type of Secret is used to store a
token credential that identifies a
{{< glossary_tooltip text="service account" term_id="service-account" >}}.

Since 1.22, this type of Secret is no longer used to mount credentials into Pods,
and obtaining tokens via the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
API is recommended instead of using service account token Secret objects.
Tokens obtained from the `TokenRequest` API are more secure than ones stored in Secret objects,
because they have a bounded lifetime and are not readable by other API clients.

> The [`kubectl create token`](/docs/reference/generated/kubectl/kubectl-commands#-em-token-em-)
> command can be used to obtain a token from the `TokenRequest` API.

Reviewer 2: ```suggestion
You can use the [`kubectl create token`](/docs/reference/generated/kubectl/kubectl-commands#-em-token-em-)
command to obtain a token from the `TokenRequest` API.
```

> If you are unable to use the `TokenRequest` API to obtain a token,
> and the security exposure of persisting a non-expiring token credential
> in a readable API object is acceptable for your use case, 
> you can create a service account token Secret object.

Reviewer 2: ```suggestion
You should only create a service account token Secret object if you can't use the `TokenRequest` API to obtain a token,
and the security exposure of persisting a non-expiring token credential
in a readable API object is acceptable to you. 
```

When using this Secret type, you need to ensure that the
`kubernetes.io/service-account.name` annotation is set to an existing
service account name. If you are creating both the ServiceAccount and

> the Secret objects, that means you should create the ServiceAccount first.

Reviewer 2: ```suggestion
the Secret objects, you should create the ServiceAccount first.
```

> the Secret objects, that means you should create the ServiceAccount first.

Reviewer 2: ```suggestion
the Secret objects, that means you should create the ServiceAccount object first.
```

After the Secret is created, a Kubernetes {{< glossary_tooltip text="controller" term_id="controller" >}}
fills in some other fields such as the `kubernetes.io/service-account.uid` annotation, and the

> `token` key in the `data` field, which is set to contain an authentication token.

Reviewer 2: What is the significance of the "which is set to"? Can you simplify to "which contains an authn token"?

Author: No significance, was pre-existing (this PR just adjusted line-breaks. Sure, rephrased.

  extra: YmFyCg==
```

> When creating a `Pod`, Kubernetes automatically finds or creates a service account

Reviewer: It doesn’t automatically create a ServiceAccount any more?

Author: The Pod bits just aren't relevant to the docs of this secret type any more

Reviewer: Fair enough.

> After creating the Secret, wait for the `token` key in the `data` field to be populated.

Reviewer: Out of curiosity, due to my other work in recognizing reconciliation: Does the secret have a ready condition or at least observedGeneration that gets set after the token is populated by the controller?

Author: No, you can tell the token is populated because the token field is set ;-)

There's no status, conditions, or observedGeneration fields in the Secret API

Reviewer: That’s unfortunate. I’m gonna have to code an exception handler for Secrets that checks for the annotation to know to wait for the token before it’s considered reconciled.

Author: Not sure I understand what you mean… the auto-generated secrets have the annotation as well. The thing to wait for is the presence of the token

Reviewer: cli-utils wasn't waiting for the Secret tokens before, probably because no one was applying them directly. So no one filed a bug report or feature request.

When a Secret is auto-generated from a ServiceAccount, the applier doesn't know that (without special business logic). It just knows to apply the ServiceAccount and wait for the ServiceAccount to be ready. But the ServiceAccount doesn't have status either. So it was considered ready immediately (which I now know is technically incorrect, because the controller takes action after seeing an update).

In legacy mode, is there any indication on the ServiceAccount that the Secret has been successfully generated?

Author: > In legacy mode, is there any indication on the ServiceAccount that the Secret has been successfully generated?

The names of any autogenerated token secrets were added to the serviceaccount's `.secrets` list

Reviewer: Of course, duh. Thanks. I'll need to add that to kstatus as an exception too.

Reviewer: Actually, I can't now, without knowing if the feature flag is enabled... Is there any way to look that up programmatically from the API?

Reviewer: This is where a status condition would be helpful. Then it could be marked as ready either way, by the controller.

Author: Let sync up offline… I'm not sure kstatus (or anything else) should be waiting on the .secrets field… I'm not quite sure what you're trying to accomplish there.

  extra: YmFyCg==
```

> After creating the Secret, wait for the `token` key in the `data` field to be populated.

Reviewer 2: ```suggestion
After creating the Secret, wait for Kubernetes to populate the `token` key in the `data` field. 
```
